{% extends 'base.html' %}

{% block body %}
{% call macros.hero(title = "My Work with the Penn-CHOP Microbiome Program", subtitle = "Hereunder is a list of some of the more presentable (and open source) projects I've worked on as a developer with the Penn-CHOP Microbiome Program. Use the buttons on each card to see my contributions. See also the new <a href='https://www.med.upenn.edu/microbiocode/' target='_blank' class='font-medium text-blue-600 underline dark:text-blue-500 hover:no-underline'>MicroBioCode</a> website!") %}{% endcall %}

{{ macros.section_header(title = "Metagenomics Pipelines") }}

<!-- Project Cards -->
<div class="container mx-auto flex flex-wrap mt-8 p-4 justify-center">
    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sunbeam", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "Sunbeam: a robust, extensible metagenomic sequencing pipeline", body = "A pipeline written in snakemake that simplifies and automates many of the steps in metagenomic sequencing analysis. It uses conda to manage dependencies and can be deployed on most Linux workstations and clusters. To read more, check out the paper in Microbiome.", contributions_link = "https://github.com/sunbeam-labs/sunbeam/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_assembly", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "Assembly, annotation, and coverage extension", body = "A Sunbeam extension for assembly of contigs using Megahit, gene annotation using Prodigal, and annotation using Blast and Diamond. It can also map reads to contigs and calculat per-base coverage using Minimap2 and samtools.", contributions_link = "https://github.com/sunbeam-labs/sbx_assembly/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_demic", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "Bacterial growth estimation extension", body = "A Sunbeam extension for estimating bacterial growth rates via peak-to-trough ratios (PTRs) using DEMIC. In preparation to use demic, reads are first assembled using Megahit, then binned by inferred genome using MaxBin2, after which reads are mapped back onto contigs using Bowtie2 and Samtools.", contributions_link = "https://github.com/sunbeam-labs/sbx_demic/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_virus_id", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "Virus identification extension", body = "A Sunbeam extension that identifies, filters, and annotates viruses from metagenomic samples. It can be configured to use multiple different assemblers and virus identification software.", contributions_link = "https://github.com/sunbeam-labs/sbx_virus_id/graphs/contributors") }}
</div>

{{ macros.section_header(title = "Statistical Packages") }}

<div class="container mx-auto flex flex-wrap mt-8 p-4 justify-center">
    {{ macros.project_card(main_link = "https://github.com/Ulthran/demic", link_opts = "target='_blank'", img_link = cdn_url + "/images/DEMIC_replication_nbg.png", title = "DEMIC: dynamic estimator of bacterial groth rates", body = "A dynamic estimator of microbial communities. It employs a multi-sample algorithm based on contigs and coverage values to infer the relative distances of contigs from the replication origin and to accurately compare bacterial growth rates between samples.", contributions_link = "https://github.com/Ulthran/demic/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/PennChopMicrobiomeProgram/ZIBR", link_opts = "target='_blank'", img_link = "", title = "ZIBR: Zero-Inflated Beta Random Effects Model", body = "A two-part zero-inflated Beta regression model with random effects for testing the association between microbial abundance and clinical covariates for longitudinal microbiome data.", contributions_link = "https://github.com/PennChopMicrobiomeProgram/ZIBR/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/lakerwsl/DAFOT", link_opts = "target='_blank'", img_link = "", title = "DAFOT: Detector of Active Flow On a Tree", body = "The main goal of this package is to provide a new method for two-sample testing for microbial compositional data by leveraging the phylogenetic tree information. Empirical evidence from real data sets suggests that the phylogenetic microbial composition difference between two populations is usually  sparse. Motivated by this observation, this package implements a new maximum type test that is particularly powerful against sparse phylogenetic composition  differences and enjoys certain optimality.", contributions_link = "https://github.com/lakerwsl/DAFOT/graphs/contributors") }}
</div>

{{ macros.section_header(title = "Utility Libraries") }}

<div class="container mx-auto flex flex-wrap mt-8 p-4 justify-center">
    {{ macros.project_card(main_link = "https://github.com/Ulthran/ShotgunUnifrac", link_opts = "target='_blank'", img_link = cdn_url + "/images/shotgun_unifrac_diagram.png", title = "ShotgunUniFrac", body = "A dual use program for downloading and extracting genes from NCBI and for creating phylogenetic trees for many marker genes and merging the results into one. This was my first from-scratch project for the PCMP.", contributions_link = "https://github.com/Ulthran/ShotgunUnifrac/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/Ulthran/pycov3", link_opts = "target='_blank'", img_link = cdn_url + "/images/pycov3_logo.png", title = "PyCov3", body = "A package for generating cov3 files which are generated from sam files giving coverage information and a fasta file giving binned contigs. Cov3 files are used as input for the DEMIC R package which calculates PTR, an estimate for bacterial growth rates.", contributions_link = "https://github.com/Ulthran/pycov3/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/PennChopMicrobiomeProgram/primertrim", link_opts = "target='_blank'", img_link = "", title = "primertrim", body = "Detect short primer sequences in FASTQ reads and trim the reads accordingly.", contributions_link = "https://github.com/PennChopMicrobiomeProgram/primertrim/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/PennChopMicrobiomeProgram/dnabc", link_opts = "target='_blank'", img_link = "", title = "DNAbc", body = "Identify DNA barcodes in FASTQ data files and write demultiplexed data.", contributions_link = "https://github.com/PennChopMicrobiomeProgram/dnabc/graphs/contributors") }}
</div>

{{ macros.section_header(title = "DevOps Tools") }}

<div class="container mx-auto flex flex-wrap mt-8 p-4 justify-center">
    {{ macros.project_card(main_link = "https://github.com/Ulthran/conda_env_check", link_opts = "target='_blank'", img_link = "", title = "Conda Env Check: a Conda environment health check workflow", body = "A GitHub Action that can provide a number of health metrics for Conda environments including verifying that it is createable, verifying that any pinned versions are retrievable, and checking versions against latest and solved.", contributions_link = "https://github.com/Ulthran/conda_env_check/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_test_action", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_test_action", body = "A GitHub Action that manages setup of the core Sunbeam pipeline for testing extensions in a CI environment. It automatically handles branching, forking, and user information to always test the correct version of any given extension.", contributions_link = "https://github.com/sunbeam-labs/sbx_test_action/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_template", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_template", body = "A template repository that allows for quick and easy creation of new Sunbeam extensions. It includes workflows to automatically fill in the new extension name and organization in many files.", contributions_link = "https://github.com/sunbeam-labs/sbx_template/graphs/contributors") }}
</div>
{% endblock %}