{% extends 'base.html' %}

{% block head %}
<title>Charlie Bushman: PCMP</title>
<meta name="description" content="Charlie Bushman is a developer in statistical metagenomics with the University of Pennsylvania and Children's Hospital of Philadelphia." />
{% endblock %}

{% block body %}
{% call macros.hero(title = "My Work with the Penn-CHOP Microbiome Program", subtitle = "Hereunder is a list of some of the more presentable (and open source) projects I've worked on as a developer with the Penn-CHOP Microbiome Program (PCMP). See also the new <a href='https://www.med.upenn.edu/microbiocode/' target='_blank' class='font-medium text-blue-600 underline dark:text-blue-500 hover:no-underline'>MicroBioCode</a> website!<br /><i class='font-small'>I've contributed to all these projects to varying degrees. A few I built from scratch, others I've updated and maintain, and to some I've just made minor additions. Use the buttons on each card to see my contributions (my GH username is Ulthran).</i>") %}{% endcall %}

<details class="container mx-auto flex flex-wrap justify-center md:w-1/2">
    <summary class="text-lg text-center">More Details</summary>

    <p class="text-md text-center">The PCMP is comprised of many labs across the University of Pennsylvania and the Children's Hospital of Philadelphia, each of which has some connection to the study of human or animal microbiomes. The software needs of these labs range from high (a core of PhD bioinformatians who need performant and reliable big data pipelines as well as the flexibility to perform fine grained exploratory analyses) to medium (mostly wet lab experiments that need follow up analysis with user-friendly and reproducible pipelines) to low (mostly wet lab and clinical work, lab members who like to live in spreadsheet land but might want a few tasks automated). It's my job as the developer for the PCMP (I'm the first and so far only pure developer for the program) to create, enhance, and maintain software for the labs in the program. If you have any unanswered questions about my work with the PCMP please reach out, I love talking about this stuff.</p>
</details>

{{ macros.section_header(title = "Metagenomics Pipelines") }}

<div class="container mx-auto flex flex-wrap justify-center">
    <p class="text-lg text-center md:w-1/2">Most of the work I do with the PCMP involves getting sequencing data from one form and transforming it to another. Doing this in a way that is well documented, reproducible, and performant is critical for being able to publish the results of downstream analysis.</p>
</div>

<!-- Project Cards -->
<div class="container mx-auto flex flex-wrap p-4 justify-center">
    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sunbeam", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "Sunbeam: a robust, extensible metagenomic sequencing pipeline", body = "A pipeline written in snakemake that simplifies and automates many of the steps in metagenomic sequencing analysis. It uses conda and docker to manage dependencies and can be deployed on most Linux workstations and clusters. It is easily extensible to perform further analyses. To read more, check out the paper in Microbiome.", contributions_link = "https://github.com/sunbeam-labs/sunbeam/graphs/contributors") }}

    {{ macros.project_card(main_link = "", link_opts = "target='_blank'", img_link = cdn_url + "/images/autobfx_logo.png", title = "AutoBfx: An automation framework for metagenomic analysis", body = "AutoBfx is starting life focused on a separate part of the metagenomic sequencing data's journey from sunbeam, automating the most common steps from directly off the machine through to early analysis. It is built on Prefect and uses an event driven architecture to automatically process new runs off the sequencers, interoperating with SLURM for computationally intensive jobs. It provides a central, fault-tolerant data pipeline (with dashboard) which saves the valuable time of bioinformaticians for more important tasks further down the analytical pipeline. The codebase is currently on enterprise GH.") }}

    {{ macros.project_card(main_link = "https://github.com/Ulthran/autobfx", link_opts = "target='_blank'", img_link = cdn_url + "/images/autobfx_logo.png", title = "Autobfx: A modern, extensible metagenomic analysis pipeline", body = "You're not seeing double, I started an offshoot of the above Autobfx to tackle the backend of the metagenomic analysis process. Like sunbeam, this tool is extensible, reproducible, and simple to use. Unlike sunbeam, it solves some of the main frustrations with the strictness of snakemake while not sacrificing on reproducibility or reusability. Autobfx allows users to run pre-made pipelines, define their own using existing components, or create plugins to integrate new tools to the ecosystem. It can interoperate with numerous compute and storage options, provides an interactive dashboard for monitoring jobs, and is installable with pip.") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_assembly", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_assembly", body = "A Sunbeam extension for assembly of contigs using Megahit, gene annotation using Prodigal, and annotation using Blast and Diamond. It can also map reads to contigs and calculat per-base coverage using Minimap2 and samtools.", contributions_link = "https://github.com/sunbeam-labs/sbx_assembly/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_coassembly", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_coassembly", body = "A Sunbeam extension to perform co-assembly of reads from arbitrary groups of samples from a given project using Megahit.", contributions_link = "https://github.com/sunbeam-labs/sbx_coassembly/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_mapping", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_mapping", body = "A Sunbeam extension for mapping reads against reference genomes using bwa and performing custom filtering.", contributions_link = "https://github.com/sunbeam-labs/sbx_mapping/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_kraken", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_kraken", body = "A Sunbeam extension for taxonomic assignment of reads to databases using Kraken. You can get pre-built kraken databases at the Kraken homepage or build your own then specify the path to your database in the config.", contributions_link = "https://github.com/sunbeam-labs/sbx_kraken/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_demic", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_demic", body = "A Sunbeam extension for estimating bacterial growth rates via peak-to-trough ratios (PTRs) using DEMIC. In preparation to use demic, reads are first assembled using Megahit, then binned by inferred genome using MaxBin2, after which reads are mapped back onto contigs using Bowtie2 and Samtools.", contributions_link = "https://github.com/sunbeam-labs/sbx_demic/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_gene_clusters", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_gene_clusters", body = "A Sunbeam extension for reads-level alignment to gene clusters of interest, e.g. bai operon or butyrate producing genes.", contributions_link = "https://github.com/sunbeam-labs/sbx_gene_clusters/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_genome_assembly", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_genome_assembly", body = "A Sunbeam extension for de novo microbial genome assembly. This pipeline uses SPAdes for single genome assembly and CheckM and read map coverage of the assembled genome for quality assessment (sbx_SPARC rules). In addition, this pipeline uses hmmer to identify SCCGs (sbx_SCCG rules).", contributions_link = "https://github.com/sunbeam-labs/sbx_genome_assembly/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_virus_id", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_virus_id", body = "A Sunbeam extension that identifies, filters, and annotates viruses from metagenomic samples. It can be configured to use multiple different assemblers and virus identification softwares.", contributions_link = "https://github.com/sunbeam-labs/sbx_virus_id/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/Ulthran/sbx_marker_magu", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_marker_magu", body = "A Sunbeam extension for detecting and quantifying phages, bacteria, and archaea in whole genome shotgun reads from human-derived samples using Marker-MAGu.", contributions_link = "https://github.com/Ulthran/sbx_marker_magu/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/Ulthran/sbx_mgv", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_mgv", body = "A Sunbeam extension for classifying viral sequences using MGV.", contributions_link = "https://github.com/Ulthran/sbx_mgv/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/Ulthran/sbx_seeker", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_seeker", body = "A Sunbeam extension for discriminating between virus and phage sequences using the alignment-free Seeker deep learning algorithm.", contributions_link = "https://github.com/Ulthran/sbx_seeker/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/Ulthran/q2-unassigner", link_opts = "target='_blank'", img_link = cdn_url + "/images/qiime2_logo.png", title = "q2-unassigner", body = "A QIIME2 plugin for evaluating the closeness of 16S rRNA marker gene sequences to named bacterial species using unassigner.", contributions_link = "https://github.com/Ulthran/q2-unassigner/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/PennChopMicrobiomeProgram/wfrcwf", link_opts = "target='_blank'", img_link = cdn_url + "/images/pcmp_logo.png", title = "WFRCWF", body = "Still in the early stages of development, WFRCWF is a pipeline DSL that will hopefully fix some of our biggest pet peeves with Snakemake and Nextflow. The basic concept is to adopt Nextflow's DAG building paradigm (explicit declaration; no solvers) but simplify it, connect it more strongly to the filesystem, and write it in python (python and R are by far more popular in the bioinformatic space than Java and its derivatives). NOTE: See Autobfx above for the most recent developments on this front.", contributions_link = "https://github.com/PennChopMicrobiomeProgram/wfrcwf/graphs/contributors") }}
</div>

{{ macros.section_header(title = "Statistical Packages") }}

<div class="container mx-auto flex flex-wrap justify-center">
    <p class="text-lg text-center md:w-1/2">Some of the cleanest and most straightforward work I do is in helping researchers developing a new statistical method to create a well-packaged software solution to accompany publication. Easy to use and install statistical packages help drive citations for the paper.</p>
</div>

<div class="container mx-auto flex flex-wrap mt-8 p-4 justify-center">
    {{ macros.project_card(main_link = "https://github.com/Ulthran/demic", link_opts = "target='_blank'", img_link = cdn_url + "/images/DEMIC_replication_nbg.png", title = "DEMIC: dynamic estimator of bacterial groth rates", body = "A dynamic estimator of microbial communities. It employs a multi-sample algorithm based on contigs and coverage values to infer the relative distances of contigs from the replication origin and to accurately compare bacterial growth rates between samples.", contributions_link = "https://github.com/Ulthran/demic/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/PennChopMicrobiomeProgram/unassigner", link_opts = "target='_blank'", title = "unassigner", body = "Evaluates consistency with named bacterial species for short 16S rRNA marker gene sequences. This is achieved by ruling out taxonmic bins for each input 16S sequence rather than by trying to assign them to specific bins, avoiding certain pitfalls of the other method such as disparate matching bins.", contributions_link = "https://github.com/PennChopMicrobiomeProgram/unassigner/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/PennChopMicrobiomeProgram/ZIBR", link_opts = "target='_blank'", img_link = "", title = "ZIBR: Zero-Inflated Beta Random Effects Model", body = "A two-part zero-inflated Beta regression model with random effects for testing the association between microbial abundance and clinical covariates for longitudinal microbiome data.", contributions_link = "https://github.com/PennChopMicrobiomeProgram/ZIBR/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/lakerwsl/DAFOT", link_opts = "target='_blank'", img_link = cdn_url + "/images/DAFOT.png", title = "DAFOT: Detector of Active Flow On a Tree", body = "The main goal of this package is to provide a new method for two-sample testing for microbial compositional data by leveraging the phylogenetic tree information. Empirical evidence from real data sets suggests that the phylogenetic microbial composition difference between two populations is usually  sparse. Motivated by this observation, this package implements a new maximum type test that is particularly powerful against sparse phylogenetic composition  differences and enjoys certain optimality.", contributions_link = "https://github.com/lakerwsl/DAFOT/graphs/contributors") }}
</div>

{{ macros.section_header(title = "Utility Libraries") }}

<div class="container mx-auto flex flex-wrap justify-center">
    <p class="text-lg text-center md:w-1/2">Some methods just aren't worth publishing, whether they're too mundane or too unique to bother. But these methods still need to be reliable and easy to use internally.</p>
</div>

<div class="container mx-auto flex flex-wrap mt-8 p-4 justify-center">
    {{ macros.project_card(main_link = "https://github.com/Ulthran/ShotgunUnifrac", link_opts = "target='_blank'", img_link = cdn_url + "/images/shotgun_unifrac_diagram.png", title = "ShotgunUniFrac", body = "A dual use program for downloading and extracting genes from NCBI and for creating phylogenetic trees for many marker genes and merging the results into one. This was my first from-scratch project for the PCMP.", contributions_link = "https://github.com/Ulthran/ShotgunUnifrac/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/Ulthran/pycov3", link_opts = "target='_blank'", img_link = cdn_url + "/images/pycov3_logo.png", title = "PyCov3", body = "A package for generating cov3 files which are generated from sam files giving coverage information and a fasta file giving binned contigs. Cov3 files are used as input for the DEMIC R package which calculates PTR, an estimate for bacterial growth rates.", contributions_link = "https://github.com/Ulthran/pycov3/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/PennChopMicrobiomeProgram/primertrim", link_opts = "target='_blank'", img_link = "", title = "primertrim", body = "Detect short primer sequences in FASTQ reads and trim the reads accordingly.", contributions_link = "https://github.com/PennChopMicrobiomeProgram/primertrim/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/PennChopMicrobiomeProgram/dnabc", link_opts = "target='_blank'", img_link = "", title = "DNAbc", body = "Identify DNA barcodes in FASTQ data files and write demultiplexed data.", contributions_link = "https://github.com/PennChopMicrobiomeProgram/dnabc/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/kylebittinger/heyfastq", link_opts = "target='_blank'", img_link = "", title = "heyfastq", body = "A Python library for reading, writing, and applying common mutations to FASTQ files.", contributions_link = "https://github.com/kylebittinger/heyfastq/graphs/contributors") }}
</div>

{{ macros.section_header(title = "Services") }}

<div class="container mx-auto flex flex-wrap justify-center">
    <p class="text-lg text-center md:w-1/2">I've built and modified some services for tracking and storing sequencing data as it passes through the PCMP.</p>
</div>

<div class="container mx-auto flex flex-wrap mt-8 p-4 justify-center">
    {{ macros.project_card(main_link = "https://github.com/PennChopMicrobiomeProgram/CHOP_metadata_checker", link_opts = "target='_blank'", img_link = cdn_url + "/images/CHOP_metadata_checker.png", title = "PCMP Metadata Checker", body = "A Flask app used by bioinformaticians and collaborators to ensure their metadata meet the standards of the PCMP. It maintains a database of projects and their metadata sheet(s). I have experimented with numerous methods for deployment for this app including on PythonAnywhere, on a dedicated CHOP server with nginx, as a serverless app on AWS (similar to this site), and, the current method, as a containerized microservice in a CHOP hosted kubernetes cluster. Unfortunately for you it is hosted on Enterprise GH to take advantage of hosted CI/CD runners and its domain is restricted to CHOP systems. But you can easily run a dev version with a test database, instructions in the readme.", contributions_link = "https://github.com/PennChopMicrobiomeProgram/CHOP_metadata_checker/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/PennChopMicrobiomeProgram/SampleRegistry", link_opts = "target='_blank'", img_link = cdn_url + "/images/sample_registry_logo.png", title = "PCMP Sample Registry", body = "A Flask app used by the CHOP Microbiome Center to track every sample they sequence. A Python library enables various mutations on the registry while the app provides data views with compiled statistics. It is deployed to a Kubernetes cluster and interacts with a CHOP hosted PostgreSQL database via SQLAlchemy. The actual site is only available on CHOP's intranet but instructions for setting up your own demo site and database are provided in the README.", contributions_link = "https://github.com/PennChopMicrobiomeProgram/SampleRegistry/graphs/contributors") }}
</div>

{{ macros.section_header(title = "DevOps Tools") }}

<div class="container mx-auto flex flex-wrap justify-center">
    <p class="text-lg text-center md:w-1/2">I've developed some tools for improving bioinformatics software development. A lot revolves around the Sunbeam ecosystem for new extensions and testing.</p>
</div>

<div class="container mx-auto flex flex-wrap mt-8 p-4 justify-center">
    {{ macros.project_card(main_link = "https://github.com/Ulthran/conda_env_check", link_opts = "target='_blank'", img_link = "", title = "Conda Env Check: a Conda environment health check workflow", body = "A GitHub Action that can provide a number of health metrics for Conda environments including verifying that it is createable, verifying that any pinned versions are retrievable, and checking versions against latest and solved.", contributions_link = "https://github.com/Ulthran/conda_env_check/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_test_action", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_test_action", body = "A GitHub Action that manages setup of the core Sunbeam pipeline for testing extensions in a CI environment. It automatically handles branching, forking, and user information to always test the correct version of any given extension.", contributions_link = "https://github.com/sunbeam-labs/sbx_test_action/graphs/contributors") }}

    {{ macros.project_card(main_link = "https://github.com/sunbeam-labs/sbx_template", link_opts = "target='_blank'", img_link = cdn_url + "/images/sunbeam_logo.png", title = "sbx_template", body = "A template repository that allows for quick and easy creation of new Sunbeam extensions. It includes workflows to automatically fill in the new extension name and organization in many files.", contributions_link = "https://github.com/sunbeam-labs/sbx_template/graphs/contributors") }}
</div>
{% endblock %}